Verify that WebGL's implicit clears obey the rasterizer discard state.

kenrussell opened the pull request:
Regression test for http://crbug.com/1149176 .

kenrussell commented:
@jdarpinian or @shrekshao or @jdashg : please review.

Test passes on Firefox; fails on Chrome and Safari. Patches incoming for both.

shrekshao reviewed (approved):
LGTM. (learn webgl implicit clearing from greggman's answer: https://stackoverflow.com/questions/54947518/requestanimationframe-readpixel-and-implicit-clear)

shrekshao commented:
(If a readpixels test needs adding as well together with drawArrays, as James points out in the CL review)

jdashg commented on sdk/tests/conformance2/rendering/rasterizer-discard-and-implicit-clear.html:
```diff
@@ -0,0 +1,104 @@
+const green = [ 0.0, 1.0, 0.0, 1.0 ];
+const transparentBlackRender = [ 0, 0, 0, 0 ];
+const greenRender = [ 0, 255, 0, 255 ];
+const sz = 256; // Must equal canvas width/height
```
const sz = gl.canvas.width = gl.canvas.height = 256;

kenrussell replied:
Thanks, fixed.

jdashg commented on sdk/tests/conformance2/rendering/rasterizer-discard-and-implicit-clear.html:
```diff
@@ -0,0 +1,104 @@
+
+const wtu = WebGLTestUtils;
+const gl = wtu.create3DContext("example", undefined, 2);
+let leftRectBuffer;
```
Doesn't need to be a global

kenrussell replied:
Right - in an earlier iteration it needed to be because the test switched between two buffers. Fixed.

jdashg commented on sdk/tests/conformance2/rendering/rasterizer-discard-and-implicit-clear.html:
```diff
@@ -0,0 +1,104 @@
+const sz = 256; // Must equal canvas width/height
+
+if (!gl) {
+    testFailed("WebGL context creation failed");
```
finishTest not called?

kenrussell replied:
Fixed.

jdashg commented on sdk/tests/conformance2/rendering/rasterizer-discard-and-implicit-clear.html:
```diff
@@ -0,0 +1,104 @@
+    gl.drawArrays(gl.TRIANGLES, 0, 6);
+    gl.disable(gl.RASTERIZER_DISCARD);
+
+    gl.uniform1f(xTranslationLoc, (30.0 - numFrames) / 30.0);
```
Would be better as `(NUM_FRAMES - framesToGo) / NUM_FRAMES`?

kenrussell replied:
Good suggestion. Fixed.

jdashg commented on sdk/tests/conformance2/rendering/rasterizer-discard-and-implicit-clear.html:
```diff
@@ -0,0 +1,104 @@
+    gl.drawArrays(gl.TRIANGLES, 0, 6);
+
+    if (numFrames-- == 0) {
+        wtu.checkCanvasRect(gl, 0, 0, sz / 2, sz, transparentBlackRender, undefined, 3);
```
Why sz/2? Comment please!

kenrussell replied:
Wrote extensive comments throughout.

jdashg reviewed (approved):
It's a test so it's fine, but I do have some nits.

kenrussell commented:
Thanks everyone for your comments. Addressed @jdashg 's review feedback. Added a test of implicit clears, rasterizer discard and readPixels based on @jdarpinian 's review feedback on the Chromium review request https://chromium-review.googlesource.com/2542770 .

The Travis CI system is slow for some reason right now. Squashing and merging this in order to get it in sooner.
